cumulus/parachains/common/src/xcm_config.rs

```diff
@@ -78,3 +86,194 @@ impl<Location: Get<MultiLocation>> ContainsPair<MultiAsset, MultiLocation>
+				return true
+			}
+		}
+		false
+	}
+}
+
+/// `MatchesLocation` implementation which works with `InteriorMultiLocation`.
+impl MatchesLocation<InteriorMultiLocation> for LocationFilter<InteriorMultiLocation> {
+	fn matches(&self, location: &InteriorMultiLocation) -> bool {
+		for filter in &self.equals_any {
+			if location.eq(filter) {
+				return true
+			}
+		}
+		for filter in &self.starts_with_any {
+			if location.starts_with(filter) {
+				return true
+			}
+		}
+		false
+	}
+}
+
+/// `FilteredNetworkExportTable` is adapter for `ExporterFor` implementation
+/// which tries to find (`bridge_location`, `fees`) for requested `network` and `remote_location`.
+///
+/// Inspired by `xcm_builder::NetworkExportTable`:
+/// the main difference is that `NetworkExportTable` does not check `remote_location`.
+pub struct FilteredNetworkExportTable<T>(sp_std::marker::PhantomData<T>);
+impl<
+		T: Get<
+			sp_std::vec::Vec<(
+				NetworkId,
+				LocationFilter<InteriorMultiLocation>,
+				MultiLocation,
+				Option<MultiAsset>,
+			)>,
+		>,
+	> ExporterFor for FilteredNetworkExportTable<T>
+{
+	fn exporter_for(
+		network: &NetworkId,
+		remote_location: &InteriorMultiLocation,
+		_: &Xcm<()>,
+	) -> Option<(MultiLocation, Option<MultiAsset>)> {
+		T::get()
+			.into_iter()
+			.find(|(ref j, location_filter, ..)| {
+				j == network && location_filter.matches(remote_location)
+			})
+			.map(|(_, _, bridge_location, p)| (bridge_location, p))
+	}
+}
+
+/// Allows execution from `origin` if it is contained in `AllowedOrigin`
+/// and if it is just a straight `Transact` which contains `AllowedCall`.
+pub struct AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>(
+	sp_std::marker::PhantomData<(RuntimeCall, AllowedCall, AllowedOrigin)>,
+);
+impl<
+		RuntimeCall: Decode,
+		AllowedCall: Contains<RuntimeCall>,
+		AllowedOrigin: Contains<MultiLocation>,
+	> ShouldExecute for AllowUnpaidTransactsFrom<RuntimeCall, AllowedCall, AllowedOrigin>
+{
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		max_weight: Weight,
+		_properties: &mut xcm_executor::traits::Properties,
+	) -> Result<(), ProcessMessageError> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowUnpaidTransactFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, properties: {:?}",
+			origin, instructions, max_weight, _properties,
+		);
+
+		// we only allow from configured origins
+		ensure!(AllowedOrigin::contains(origin), ProcessMessageError::Unsupported);
+
+		// we expect an XCM program with single `Transact` call
+		instructions
+			.matcher()
+			.assert_remaining_insts(1)?
+			.match_next_inst(|inst| match inst {
+				Transact { origin_kind: OriginKind::Xcm, call: encoded_call, .. } => {
+					// this is a hack - don't know if there's a way to do that properly
+					// or else we can simply allow all calls
+					let mut decoded_call = DoubleEncoded::<RuntimeCall>::from(encoded_call.clone());
```

Comment:
I just cleaned it little bit here: https://github.com/paritytech/polkadot-sdk/pull/1215/commits/791b88a84a154bfac4448cec3005484aae722cc0#diff-c27f7ffe369c3d6f8b46fd6d0e487a72f950e50364f8ee070f283e57ea032b46L212-R220

I dont know, I dont see much usage of `Any` or `downcast_mut` for runtimes,
but if you can make it work and looks ok, then lets change it

Reply:
The problem is that now we're cloning. Maybe downcasting is a bit complicating. But for example something like this could be ok-ish and more efficient:
```
					if std::mem::size_of::<Call>() != std::mem::size_of::<RuntimeCall>() {
						return Err(ProcessMessageError::BadFormat)
					}
					let encoded_call = unsafe {
						sp_std::mem::transmute::<_, &mut DoubleEncoded<RuntimeCall>>(encoded_call)
					};
```

instead of `encoded_call.clone().into::<RuntimeCall>()`.

Maybe it could implemented directly in a method defined on the `DoubleEncoded` struct.

Reply:
> I just cleaned it little bit here: https://github.com/paritytech/polkadot-sdk/commit/791b88a84a154bfac4448cec3005484aae722cc0#diff-c27f7ffe369c3d6f8b46fd6d0e487a72f950e50364f8ee070f283e57ea032b46L212-R220

I think this is ok, even with the `clone` - this happens on calls so it's not that often and `encoded_call` is not a huge array of bytes to make a difference in cost.

I am definitely against:
```
unsafe {
	sp_std::mem::transmute::<_, &mut DoubleEncoded<RuntimeCall>>(encoded_call)
}
```
even if it is sane now, it can possibly later fail in subtle ways, resulting in UB in runtime which'd be really bad.

Reply:
Oh, you're right, this shouldn't happen very often. Yes, than cloning should be ok.